internal/lib/control_config.go

```diff
@@ -0,0 +1,236 @@
+	akoinformer "github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/pkg/client/v1alpha1/informers/externalversions/ako/v1alpha1"
+	"github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/pkg/utils"
+	advl4crd "github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/third_party/service-apis/client/clientset/versioned"
+	advl4informer "github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/third_party/service-apis/client/informers/externalversions/apis/v1alpha1pre1"
+)
+
+type AdvL4Informers struct {
+	GatewayInformer      advl4informer.GatewayInformer
+	GatewayClassInformer advl4informer.GatewayClassInformer
+}
+
+type ServicesAPIInformers struct {
+	GatewayInformer      svcInformer.GatewayInformer
+	GatewayClassInformer svcInformer.GatewayClassInformer
+}
+
+type AKOCrdInformers struct {
+	HostRuleInformer        akoinformer.HostRuleInformer
+	HTTPRuleInformer        akoinformer.HTTPRuleInformer
+	AviInfraSettingInformer akoinformer.AviInfraSettingInformer
+}
+
+type IstioCRDInformers struct {
+	VirtualServiceInformer  istioInformer.VirtualServiceInformer
+	DestinationRuleInformer istioInformer.DestinationRuleInformer
+	GatewayInformer         istioInformer.GatewayInformer
+}
+
+// akoControlConfig struct is intended to store all AKO related global
+// variables, that are set as part of AKO bootup. This is a store of client-sets,
+// informers, config parameters, and internally computed static configurations.
+// TODO (shchauhan): Add other global parameters, which are currently present as independent
+// global variables as part of lib.go
+type akoControlConfig struct {
+	// client-set and informer for v1alpha1pre1 services API.
+	advL4Clientset    advl4crd.Interface
+	akoAdvL4Informers *AdvL4Informers
+
+	// client-set and informer for v1alpha1 services API.
+	svcAPICS        svcapi.Interface
+	svcAPIInformers *ServicesAPIInformers
+
+	// client-set and informer for v1alpha1 AKO CRDs.
+	crdClientset akocrd.Interface
+	crdInformers *AKOCrdInformers
+
+	// client-set and informer for v1alpha3 istio CRDs.
+	istioClientset istiocrd.Interface
+	istioInformers *IstioCRDInformers
+
+	// akoEventRecorder is used to store record.akoEventRecorder
+	// that allows AKO to broadcast kubernetes Events.
+	akoEventRecorder *utils.EventRecorder
+
+	// akoPodObjectMeta holds AKO Pod ObjectMeta information
+	akoPodObjectMeta *metav1.ObjectMeta
+
+	// aviInfraSettingEnabled is set to true if the cluster has
+	// AviInfraSetting CRD installed.
+	aviInfraSettingEnabled bool
+	// hostRuleEnabled is set to true if the cluster has
+	// HostRule CRD installed.
+	hostRuleEnabled bool
+	// httpRuleEnabled is set to true if the cluster has
+	// HTTPRule CRD installed.
+	httpRuleEnabled bool
+}
+
+var akoControlConfigInstance *akoControlConfig
+
+func AKOControlConfig() *akoControlConfig {
+	if akoControlConfigInstance == nil {
```

Comment:
I think it will be good if we create singleton object of akoControlConfig  and use it over all place instead of calling lib.AKOControlConfig each time.

Reply:
that is what it is @akshayhavile. `lib.AKOControlConfig()` is a function to access the singleton. I initialize it only if it is nil.

Reply:
This code doesn't seem to be thread safe. If multiple threads(in case) try to access this code, same variable might get initialised twice. But i think impact will be less.

Reply:
I could add a initialize Once block for it, but this gets initilaized in the AKO main function, so getting initialized in multiple goroutines should not happen. Also since it is a global struct, we donot expect it to be initialized from different goroutines.